### ISSUE TITLE
Render JSON using a template renderer?

### ISSUE THREAD

**Issue by bakura10** [zendframework/zend-expressive#182]
Hi :),

I've been trying to install a template library (in this case, Plates) in order to use it for rendering JSON (so I can separate my controller from my rendering code).

My middleware's invoke method therefore does something like that:

```php
return new JsonResponse($this->template->render('app::my-resource', ['resource' => $object]));
```

My template code is a simple file that looks like that:

```php
return [
  'foo' => $object->getFoo()
]
```

I verified and the template is properly called, so my configuration is right. However the fact that it extract the content using ob_start makes me think that I'm using it the wrong way.

What is the recommended way to achieve that? I'm liking the idea of rendering my resources in a separate file, so I can have access to plates features, as well as being able to render URLs and so on...

Thanks!

**Comment by moderndeveloperllc** [comment]
Have you thought about perhaps having a Rendering.php middleware that is called in the `post_routing` part of the configuration? That would keep the templates out of the routed middleware. It also helps keep to the middleware-as-service model as opposed to the middleware-as-controller way of thinking about it.

**Comment by bakura10** [comment]
I didn't try that, but what I want is actually having the few useful features of Plates and it's resolving mechanism. What do you suggest exactly by using a post_routing middleware?
Also, what do you mean by "middleware-as-service" actually?

**Comment by moderndeveloperllc** [comment]
Sorry for the confusion. Looking over your question again, the documentation, and what you are trying to do: I think that you are doing it the recommended way. @weierophinney?

The reason it seems odd is that these templating solutions are designed to "do it all" and be the last step in a standard MVC setup. This means the libraries are designed to create and return the entire response - header, body, etc. Since Expressive utilizes the PSR-7-compliant Diactoros library to compose the response, all that the templating engines are being asked to do is just compose the *body* of the response to be injected into a new Diactoros response object. I guess Plates does this by reading its output with `ob_start()`/`ob_end()`. The phrase "middleware-as-service" is something I use to remind myself that a middleware is supposed to do only do one thing as opposed to having lots of actions and handle lots of stuff like an MVC controller. I misread your question and thought that you were modifying the response for all middleware before passing it to a template. In that case, I would just pass JSON out of the routed middleware and have a `post_routing` middleware do the needed modifications to keep things DRY.

**tl;dr**: You are grabbing the the composed body from the middle of the template library processing chain - it can seem a little messy internally.

**Comment by danizord** [comment]
@bakura10 That will not work with `JsonReponse` because [TemplateRendererInterface#render()](https://github.com/zendframework/zend-expressive/blob/master/src/Template/TemplateRendererInterface.php#L43) returns `string` and `JsonResponse` constructor expects `array`.

However you can achieve your needs this way:
```php
# Template file: my-resource.php
echo json_encode([
    'foo' => $object->getFoo()
]);
```

```php
# Middleware
$response->getBody()->write(
    $this->template->render('app::my-resource', ['resource' => $object])
);

return $response->withHeader('content-type', 'application/json')
```

**Comment by bakura10** [comment]
Yes exactly, that's the only solution, but I'd like to avoid the boilerplate of doing that, and I don't really like the echo stuff. I suppose I have no other choice than creating my own repsonse. But maybe JsonResponse should not try to encode if the content is alreayd a string (@weierophinney thoughts?)

**Comment by weierophinney** [comment]
@bakura10 I'd create a `JsonStringResponse` that accepts an already encoded JSON string, but then sets the appropriate headers, personally — or just manipulate the response provided to you (as that eliminates a dependency on a concrete response implementation).

**Comment by grizzm0** [comment]
@bakura10 Any progress on this? After zfrrest 0.4.0 I'd also like to render my json in a template file. ;)

**Comment by bakura10** [comment]
Haha no not yet. But creating a custom response seems like a good idea. I'll see if I cannot tweak the JsonResponse so that when it detects it is a string, it does not serialise again.

**Comment by weierophinney** [comment]
I'm closing this, as the solution will be a custom response type or changes to existing response types, both of which are out of scope for this repository.
